ugc/storage.cpp

```diff
@@ -1,33 +1,434 @@
+  SerializerJson<Sink> ser(sink);
+  ser(update);
+
+  my::Json serializedUgc(data.data());
+  node = serializedUgc.get();
+  auto embeddedNode = my::NewJSONObject();
+  ToJSONObject(*embeddedNode, kDataVersionKey, index.m_dataVersion);
+  ToJSONObject(*embeddedNode, kMwmNameKey, index.m_mwmName);
+  ToJSONObject(*embeddedNode, kFeatureIdKey, index.m_featureId);
+  ToJSONObject(*node, kFeatureKey, *embeddedNode);
+  embeddedNode.release();
+}
+}  // namespace
+
+Storage::Storage(Index const & index)
+  : m_index(index)
 {
   Load();
 }
 
-UGCUpdate const * Storage::GetUGCUpdate(FeatureID const & id) const
+UGCUpdate Storage::GetUGCUpdate(FeatureID const & id) const
 {
-  auto const it = m_ugc.find(id);
-  if (it != end(m_ugc))
-    return &it->second;
-  return nullptr;
+  ASSERT_THREAD_CHECKER(m_threadChecker, ());
+  if (m_UGCIndexes.empty())
+    return {};
+
+  auto const feature = GetOriginalFeature(id);
+  CHECK(feature, ());
+  auto const & mercator = feature->GetCenter();
+  feature::TypesHolder th(*feature);
+  th.SortBySpec();
+  auto const type = th.GetBestType();
+
+  auto const index = find_if(m_UGCIndexes.begin(), m_UGCIndexes.end(), [type, &mercator](UGCIndex const & index) -> bool
+  {
+    return type == index.m_type && mercator == index.m_mercator && !index.m_isDeleted;
+  });
+
+  if (index == m_UGCIndexes.end())
+    return {};
+
+  auto const offset = index->m_offset;
+  auto const nextIndex = index + 1;
+  uint64_t nextOffset;
+  if (nextIndex == m_UGCIndexes.end())
+    CHECK(GetPlatform().GetFileSizeByName(kUGCUpdateFileName, nextOffset), ());
+  else
+    nextOffset = nextIndex->m_offset;
+
+  auto const size = nextOffset - offset;
+  vector<uint8_t> buf;
+  auto const ugcFilePath = GetUGCFilePath();
+  try
+  {
+    FileReader r(ugcFilePath);
+    r.Read(offset, buf.data(), size);
+  }
+  catch (RootException const & exception)
+  {
+    LOG(LWARNING, ("Exception while reading file:", ugcFilePath));
+    return {};
+  }
+
+  MemReader r(buf.data(), buf.size());
+  NonOwningReaderSource source(r);
+  UGCUpdate update;
+  Deserialize(source, update);
+  return update;
 }
 
 void Storage::SetUGCUpdate(FeatureID const & id, UGCUpdate const & ugc)
 {
-  m_ugc[id] = ugc;
+  ASSERT_THREAD_CHECKER(m_threadChecker, ());
+  auto const feature = GetOriginalFeature(id);
+  CHECK(feature, ());
+
+  auto const & mercator = feature->GetCenter();
+  feature::TypesHolder th(*feature);
+  th.SortBySpec();
+  auto const type = th.GetBestType();
+  for (auto & index : m_UGCIndexes)
+  {
+    if (type == index.m_type && mercator == index.m_mercator && !index.m_isDeleted)
+    {
+      index.m_isDeleted = true;
+      m_numberOfDeleted++;
+      break;
+    }
+  }
+  // TODO: Call Defragmentation().
+
+  auto const ugcFilePath = GetUGCFilePath();
+  try
+  {
+    FileWriter w(ugcFilePath);
+    Serialize(w, ugc);
+  }
+  catch (RootException const & exception)
+  {
+    LOG(LWARNING, ("Exception while writing file:", ugcFilePath));
```

Comment:
This should be ERROR, not WARNING, right? And it would be nice to print exception.What().